Brokerages/Alpaca/AlpacaBrokerage.cs

```diff
@@ -0,0 +1,186 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using NodaTime;
+using QuantConnect.Data;
+using QuantConnect.Data.Market;
+using QuantConnect.Interfaces;
+using QuantConnect.Logging;
+using QuantConnect.Packets;
+using QuantConnect.Securities;
+using HistoryRequest = QuantConnect.Data.HistoryRequest;
+using Order = QuantConnect.Orders.Order;
+
+namespace QuantConnect.Brokerages.Alpaca
+{
+    /// <summary>
+    /// Alpaca Brokerage implementation
+    /// </summary>
+    [BrokerageFactory(typeof(AlpacaBrokerageFactory))]
+    public class AlpacaBrokerage : AlpacaApiBase
+    {
+        /// <summary>
+        /// The maximum number of bars per historical data request
+        /// </summary>
+        public const int MaxBarsPerRequest = 5000;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AlpacaBrokerage"/> class.
+        /// </summary>
+        /// <param name="orderProvider">The order provider.</param>
+        /// <param name="securityProvider">The holdings provider.</param>
+        /// <param name="keyId">The Alpaca api key id</param>
+        /// <param name="secretKey">The api secret key</param>
+        /// <param name="tradingMode">The Alpaca trading mode. paper/live</param>
+        public AlpacaBrokerage(IOrderProvider orderProvider, ISecurityProvider securityProvider, string keyId, string secretKey, string tradingMode)
+            : base("Alpaca Brokerage")
+        {
+            var baseUrl = "api.alpaca.markets";
+            if (tradingMode.Equals("paper")) baseUrl = "paper-" + baseUrl;
+            baseUrl = "https://" + baseUrl;
+            base.initialize(orderProvider, securityProvider, keyId, secretKey, baseUrl);
+
+        }
+
+        #region IBrokerage implementation
+
+        /// <summary>
+        /// Returns true if we're currently connected to the broker
+        /// </summary>
+        public override bool IsConnected
+        {
+            get { return IsConnected; }
```

Comment:
This causes a stack overflow exception -- please run all the tests to confirm recent changes haven't broken anytihng.

Reply:
Okay, let me check.